WSC Flow phase 1 - support for none Intel agent

arnout opened the pull request:
The M2 still needs to be filled in with the relevant information, but that's a bigger change because it requires the database to be updated with SSID information (currently SSID updates are only performed from an explicit command).

This PR goes on top of #154.

arnout committed:
controller: master: remove check for Intel manufacturer

Now we have basic support for non-Intel agents in the controller, remove
the manufacturer check in the M1 handling.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
agent: slave: remove check for Intel manufacturer

Now we have basic support for non-Intel controllers in the agent, remove
the manufacturer check in the M1 handling.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

tomereli commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -500,10 +500,98 @@ bool master_thread::handle_cmdu_1905_autoconfiguration_WSC(Socket *sd,
+            return false;
+        }
+    } else {
+        LOG(INFO) << "Non-Intel radio agent join (al_mac=" << al_mac << " ruid=" << ruid << ")";
```
I think all this should go to a separate function

tomereli commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -490,14 +490,6 @@ bool master_thread::handle_cmdu_1905_autoconfiguration_WSC(Socket *sd,
         }
     }
 
-    std::string manufacturer =
```
I think it is beneficial to still have the manufacturer at least printed out here, and also add the manufacturer to the node structure in the DB so it will be seen in the network map.

arnout replied:
There is currently nothing in the DB for this information.

If we consider adding it, then maybe we should first think about refactoring the database in general...

tomereli replied:
(nitpick) perhaps we can just print the manufacturer here so we'll have it in the log?

tomereli commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4349,15 +4349,61 @@ bool slave_thread::handle_autoconfiguration_wsc(Socket *sd, ieee1905_1::CmduMess
-        LOG(ERROR) << "Not vendor specific TLV (not Intel?)";
-        return false;
-    }
+    if (cmdu_rx.getNextTlvType() == uint8_t(ieee1905_1::eTlvType::TLV_VENDOR_SPECIFIC)) {
```
still - a few lines up the manufacturer check will return false for none intel controller.

arnout committed:
controller: master: handle non-intel M1

If the AP Autoconfiguration WSC M1 message does not have the Intel
Vendor-Specific TLV, still handle the message (and send an M2).

"Handling" just means updating the topology in the database.
For now, this still contains a lot of shortcuts and default values,
because the information in the M1 message does not correspond very well
with the information in the Intel slave join message.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout commented:
I finally understand why I was strugling so much with this PR. The problem is that the Intel slave join message really has nothing to do with 1905.1 autoconfig - it really is a topology message.

Therefore, I'm not so sure what to do with this. In the non-Intel case, we really should build up this information from the 1905.1 Topology Response message, which contains much more of the required information. It contains no information about the radios, but that should be retrieved from AP Capability Report messages, which is a separate flow. So doing all of this in the WSC M1 seems wrong to me...

arnout committed:
agent: slave: support non-Intel controller

In order to test if the controller supports a non-Intel agent, we also
need the agent to support a non-Intel controller.

Similarly to how it's done for the controller, copy some of the handling
of the Intel VS message to the handling of the WSC M2.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
controller: master: add manufacturer to node in database

Since the manufacturer is no longer always Intel, it makes sense to
store it in the node information in the database.

In the Intel case, it is just set to Intel.

In the non-Intel case, the manufacturer string from the M1 is used.
Since we create a bridge node and a slave node from the M1, both are
set.

Include the manufacturer name in the node info when it is printed.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

tomereli commented:
(nitpick) we should consider adding an option to the agent which will make it start in generic mode. In this mode it will not identify itself in autoconfig as Intel, but as "prplMesh" manufacturer.
We should remove the trigger for tasks which are intel specific in the controller for none Intel agents and add a different print to the agent to mark it as operational, maybe after autoconfig success. This can be done in a different PR of course.

Even better - we can have one radio agent be Intel manufacturer and the other prplMesh, this way we are testing both modes in CI.

arnout commented:
> (nitpick) we should consider adding an option to the agent which will make it start in generic mode. In this mode it will not identify itself in autoconfig as Intel, but as "prplMesh" manufacturer.

That is in fact an excellent idea! Do you prefer that in this PR or in a new one?

> We should remove the trigger for tasks which are intel specific in the controller for none Intel agents and add a different print to the agent to mark it as operational, maybe after autoconfig success.

I'm not sure how to do that. Could I give that to @morantr ?

tomereli commented:
> > (nitpick) we should consider adding an option to the agent which will make it start in generic mode. In this mode it will not identify itself in autoconfig as Intel, but as "prplMesh" manufacturer.
> 
> That is in fact an excellent idea! Do you prefer that in this PR or in a new one?

This PR will be complete with this - so I suggest you start as a new commit in this PR, and if you don't finish by EOD we'll move it to a different PR so we can merge this one today.

> 
> > We should remove the trigger for tasks which are intel specific in the controller for none Intel agents and add a different print to the agent to mark it as operational, maybe after autoconfig success.
> 
> I'm not sure how to do that. Could I give that to @morantr ?

Yes - Moran can start working on that after the CAPI stuff. Could you open a task for this?

tomereli commented:
> (nitpick) we should consider adding an option to the agent which will make it start in generic mode. In this mode it will not identify itself in autoconfig as Intel, but as "prplMesh" manufacturer.
> We should remove the trigger for tasks which are intel specific in the controller for none Intel agents and add a different print to the agent to mark it as operational, maybe after autoconfig success.

One small improvement to the above - we can set one radio agent to be Intel, and the other to be generic, both by default. Then we can test in CI that the 2 modes are working. The option should go to the `beerocks_agent.conf` - we add the manufacturer there and the agent will use this in the WSC autoconfig flow.

tomereli commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1425,6 +1440,92 @@ bool master_thread::handle_intel_slave_join(Socket *sd, ieee1905_1::CmduMessageR
+    if (database.has_node(radio_mac)) {
+        if (database.get_node_type(radio_mac) != beerocks::TYPE_SLAVE) {
+            database.set_node_type(radio_mac, beerocks::TYPE_SLAVE);
+            LOG(ERROR) << "Existing mac node is not TYPE_SLAVE";
```
Shouldn't we return false here?

arnout replied:
> Shouldn't we return false here?

No.

The Intel slave handling doesn't do that.

Also, it makes sense that more recent information takes precedence over whatever happens to be stored in the database - as long as that is done in a consistent way. Setting the type to TYPE_SLAVE does exactly that.

tomereli commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1425,6 +1440,92 @@ bool master_thread::handle_intel_slave_join(Socket *sd, ieee1905_1::CmduMessageR
+        database.add_node(radio_mac, bridge_mac, beerocks::TYPE_SLAVE, radio_mac);
+    }
+    database.set_hostap_is_acs_enabled(radio_mac, false);
+    // TODO
```
There should be no slave reconf in generic mode - we have WSC autoconfig renew for that, so I guess this TODO should be removed.

tomereli commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1425,6 +1440,92 @@ bool master_thread::handle_intel_slave_join(Socket *sd, ieee1905_1::CmduMessageR
+    // TODO Assume repeater mode
+    database.set_hostap_repeater_mode_flag(radio_mac, true);
+    // TODO Assume no backhaul manager
+    database.set_hostap_backhaul_manager(radio_mac, false);
```
This could be problematic since one of the agents in our implementation must be the "backhaul", so in the case of both radio agents being non-Intel the controller behaviour may be unexpected, we should verify that my concerns are invalid once we have the option to start both repeater radios in generic mode.

arnout replied:
> This could be problematic since one of the agents in our implementation must be the "backhaul", so in the case of both radio agents being non-Intel the controller behaviour may be unexpected, we should verify that my concerns are invalid once we have the option to start both repeater radios in generic mode.

I don't think so. AFAICS, in our current test, neither of the radios in the repeater is a backhaul (it's wired backhaul).

tomereli commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1425,6 +1440,92 @@ bool master_thread::handle_intel_slave_join(Socket *sd, ieee1905_1::CmduMessageR
+    // TODO assume SSIDs are not hidden
+    database.set_hostap_advertise_ssid_flag(radio_mac, true);
+
+    // TODO
```
WSC autoconfig message contains the capabilities in the WSC M1 RF_Bands attribute:
![image](https://user-images.githubusercontent.com/13386849/63024133-46546100-beaf-11e9-883b-31e9dd100657.png)
We should use this information here.

arnout replied:
> WSC autoconfig message contains the capabilities in the WSC M1 RF_Bands attribute:
> We should use this information here.

I skipped that because in the current database, this information is not kept separately. Instead, it is constructed out of the channel information. We should get the channel information from the AP Capability Report message.

tomereli commented on controller/src/beerocks/master/son_master_thread.h:
```diff
@@ -54,6 +54,9 @@ class master_thread : public beerocks::socket_thread {
     bool handle_intel_slave_join(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx,
                                  ieee1905_1::CmduMessageTx &cmdu_tx);
+    bool handle_non_intel_slave_join(Socket *sd, std::shared_ptr<ieee1905_1::tlvWscM1> tlvwscM1,
+                                     std::string bridge_mac, std::string radio_mac,
```
Better to use `const std::string` or `std::string &` to avoid copy constructor since we're not changing `bridge_mac` and `radio_mac`.

tomereli commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4397,15 +4397,62 @@ bool slave_thread::handle_autoconfiguration_wsc(Socket *sd, ieee1905_1::CmduMess
-    if (!parse_intel_join_response(sd, cmdu_rx)) {
-        LOG(ERROR) << "Parse join response failed";
-        return false;
+        // request the current vap list from ap_manager
```
Deserves a separate function IMHO

tomereli commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4397,15 +4397,62 @@ bool slave_thread::handle_autoconfiguration_wsc(Socket *sd, ieee1905_1::CmduMess
+            return false;
+        }
+    } else {
+        LOG(INFO) << "Non-Intel controller join response";
```
Could be helpful for future debugging to print the manufacturer string as part of this log print.

arnout replied:
It's already printed together with the encrypted settings.

tomereli commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4397,15 +4397,62 @@ bool slave_thread::handle_autoconfiguration_wsc(Socket *sd, ieee1905_1::CmduMess
+        }
+        pending_client_association_cmdu.clear();
+
+        // No version checking for non-Intel controller
```
Agree - so why keep the TODO?

arnout replied:
I don't understand your question. The TODO is about passing the master/slave versions to the platform manager, where it will eventually be used by bml_get_master_slave_versions. The "No version checking" comment is about not checking the version (i.e. the version is always considered correct).

tomereli commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4388,13 +4388,6 @@ bool slave_thread::handle_autoconfiguration_wsc(Socket *sd, ieee1905_1::CmduMess
                    << "     authentication_type: " << authentication_type << std::endl
                    << "     encryption_type: " << encryption_type << std::endl;
-
-        if (!manufacturer.compare("Intel")) {
```
Can we add the manufacturer print in the above log? will be useful for future debugging.

arnout replied:
It is already there.

tomereli commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -200,6 +200,25 @@ std::string db::get_node_ipv4(std::string mac)
     return n->ipv4;
 }
 
+bool db::set_node_manufacturer(std::string mac, std::string manufacturer)
```
I know the rest of this file doen't follow this - but better to have the parameters as const. This applies to all newly added functions below.

arnout replied:
You made this comment in another review as well, but I don't understand it. Do you really want a signature like

```
void foo(const int x1, const bool yesno, const std::string bar);
```

?

Or do you only want const-ness for `std::string` and not for scalar types?

Or do you actually want a const ref `const std::string&` because you don't trust the copyref optimisation?

tomereli replied:
I prefer a signature with const values only for non-scalar types whenever possible. Also the function should be declared `const` if it a class function and doesn't change any members:

```
void foo(,int x1, bool yesno, const std::string &str) const; //function const for class members when possible
```
If the parameter is inout, then it should be passed without const.
If the function returns scalar type, it should return none const (anyway copied).
If the function returns non-scalar type - then it should return const when possible to save an extra copy constructor call.

Of course all these are really (nitpicks), so don't bother changing anything if you disagree.
Also 
```

arnout commented:
> > > We should remove the trigger for tasks which are intel specific in the controller for none Intel agents and add a different print to the agent to mark it as operational, maybe after autoconfig success.
> > 
> > I'm not sure how to do that. Could I give that to @morantr ?
> 
> Yes - Moran can start working on that after the CAPI stuff. Could you open a task for this?

#162

arnout committed:
controller: master: dump M1 info when receiving WSC

To help debugging non-Intel agent join, dump their M1 info when the WSC
message is received.

Reorder the assignments a bit to make this easier.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>
